server/server.go

```diff
@@ -236,15 +256,157 @@ func (cs *channelStore) createChannel(s *StanServer, name string) (*channel, err
+			return err
+		}
+	}
+	store, err := raftboltdb.NewBoltStore(filepath.Join(path, raftLogFile))
+	if err != nil {
+		return err
+	}
+	c.raftStore = store
+	cacheStore, err := raft.NewLogCache(logCacheSize, store)
+	if err != nil {
+		store.Close()
+		return err
+	}
+
+	config := raft.DefaultConfig()
+
+	// FIXME: Send output of Raft logger
+	// Raft expects a *log.Logger so can not swap that with another one,
+	// but we can modify the outputs to which the Raft logger is writing.
+	//config.Logger = s.log
+	logReader, logWriter := io.Pipe()
+	config.LogOutput = logWriter
+	bufr := bufio.NewReader(logReader)
+	go func() {
+		for {
+			line, _, err := bufr.ReadLine()
+			if err != nil {
+				s.log.Errorf("error while reading piped output from Raft log: %s", err)
+				break
+			}
+
+			fields := bytes.Fields(line)
+			level := string(fields[2])
+			raftLogFields := fields[4:]
+			raftLog := string(bytes.Join(raftLogFields, []byte(" ")))
+
+			switch level {
+			case "[DEBUG]":
+				s.log.Tracef("%v", raftLog)
+			case "[INFO]":
+				s.log.Noticef("%v", raftLog)
+			case "[WARN]":
+				s.log.Noticef("%v", raftLog)
+			case "[ERROR]":
+				s.log.Fatalf("%v", raftLog)
+			default:
+				s.log.Noticef("%v", raftLog)
+			}
+		}
+	}()
+
+	snapshotStore, err := raft.NewFileSnapshotStore(path, numLogSnapshots, logWriter)
+	if err != nil {
+		store.Close()
+		return err
+	}
+
+	// TODO: using a single NATS conn for every channel might be a bottleneck. Maybe pool conns?
+	transport, err := natslog.NewNATSTransport(
+		fmt.Sprintf("%s.%s", s.opts.ClusterNodeID, c.name), s.ncr, 2*time.Second, logWriter)
+	if err != nil {
+		store.Close()
+		return err
+	}
+
+	peers := make([]string, len(s.opts.ClusterPeers))
+	for i, p := range s.opts.ClusterPeers {
+		peers[i] = fmt.Sprintf("%s.%s", p, c.name)
+	}
+	peersStore := &raft.StaticPeers{StaticPeers: peers}
+
+	node, err := raft.NewRaft(config, c, cacheStore, store, snapshotStore, peersStore, transport)
+	if err != nil {
+		transport.Close()
+		store.Close()
+		return err
+	}
+	c.raft = node
+
+	// Wait for a leader to be elected before proceeding.
+	deadline := time.Now().Add(30 * time.Second)
+	for time.Now().Before(deadline) {
+		if leader := node.Leader(); leader != "" {
+			// Leader was elected.
+			break
+		}
+		// Wait a bit.
+		time.Sleep(5 * time.Millisecond)
+	}
+	if node.Leader() == "" {
+		node.Shutdown()
+		transport.Close()
+		store.Close()
+		return errors.New("channel leader was not elected in time")
+	}
+
+	if node.State() == raft.Leader {
+		atomic.StoreUint32(&c.leader, 1)
+	}
+
+	go func() {
+		select {
+		case isLeader := <-node.LeaderCh():
+			if isLeader {
+				// Use a barrier to ensure all preceding operations are
+				// applied to the FSM, then update nextSequence.
+				if err := node.Barrier(0).Error(); err != nil {
+					// TODO: probably step down as leader?
+					panic(err)
+				}
+				lastSequence, err := c.store.Msgs.LastSequence()
```

Comment:
If I understand correctly, we are here if the node was a follower and now becomes leader.
It is setting its nextSequence based on what is the last sequence from the store.
We use atomic for nextSequence because in the ioLoop, we may be in a 'replicate' call
that uses this field. What if here we were posting a control message which is processed 
the clientPublish handler. We could schedule a special iopm message to the ioLoop.
when processing this, we notify a channel the code will be waiting on here.
This would allow us not to use atomics for nextSequence.
I do something similar in remove_channel branch since I need to ensure that all
pending messages need to be done with a given channel before processing the request.
I could experiment here and will let you know if it is worth it.